sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -652,14 +653,18 @@ public virtual async Task StopProcessingAsync(CancellationToken cancellationToke
         private async Task RunReceiveTaskAsync(
             CancellationToken cancellationToken)
         {
-            List<Task> tasks = new List<Task>();
+            List<Task> tasks = new List<Task>(MaxConcurrentCalls + 1);
```

Comment:
If we've got a known size, I wonder if an array would be more efficient to use with `Task.WhenAll`

Reply:
It may not be worth the effort, as we'd have to manage the overflow preemptively rather than using the approach on 673-675

Reply:
I think the array would be very tricky since you would need then to deal with tasks completed inside any position of the array and then empty those while also making sure to fill in only empty positions. I think the list approach is still good for the use case here

Reply:
I don't think a size of MaxConcurrentCalls + 1 will be enough, since we can potentially add up to MaxConcurrentCalls tasks in the foreach loop, so it should probably be 2*MaxConcurrentCalls.

Reply:
Should I assign max Calls to a field and use that instead of recalculating?

Reply:
Actually, looks like we calculate this here - https://github.com/Azure/azure-sdk-for-net/blob/f0b4c27975e6ef4371a4a45a6ec1a7a07ffc8ca2/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs#L217 (there appears to be a bug now as we are using MaxConcurrentCalls which I believe will be 0 when using sessions, but the impact of this is minor as it would just cause us to clean out the completed tasks more often than we intended). We can save this into a field, but would still need to add _receiverManagers.Count to account for the overflow from the current loop. Alternatively, we can just move the "clean out the completed tasks" stuff to before we add the new tasks - this might be simpler.

Reply:
I've pushed a commit that makes use of the calculated value. I thought about

>Alternatively, we can just move the "clean out the completed tasks" stuff to before we add the new tasks - this might be simpler.

and I was thinking that if we just do `RemoveAll(...)` we might be operating more on the list again. Let me know what you think about the current version.

Reply:
Looks good to me